From: Stained Glass OS <ke7oxh@gmail.com>
Subject: [PATCH] msxml3: version 6 properties, collection default item (2031)

The version 6 document keeps NormalizeAttributeValues (default false) and
MaxElementDepth (default 256) between setProperty and getProperty. The
IDispatchEx default item of node lists and attribute maps checks its
parameter count (DISP_E_BADPARAMCOUNT), answers an empty-collection index
with a null object, and every path through either object leaves an empty
result. Twelve passing todo_wine marks removed in the domdoc tests.
---
--- a/dlls/msxml3/domdoc.c
+++ b/dlls/msxml3/domdoc.c
@@ -77,6 +77,8 @@
     LONG selectNsStr_len;
     BOOL XPath;
     IUri *uri;
+    VARIANT_BOOL normalize_attributes;
+    LONG max_element_depth;
 } domdoc_properties;
 
 typedef struct ConnectionPoint ConnectionPoint;
@@ -289,6 +291,8 @@
     /* properties that are dependent on object versions */
     properties->version = version;
     properties->XPath = (version == MSXML4 || version == MSXML6);
+    properties->normalize_attributes = VARIANT_FALSE;
+    properties->max_element_depth = 256;
 
     /* document uri */
     properties->uri = NULL;
@@ -314,6 +318,8 @@
         if (pcopy->schemaCache)
             IXMLDOMSchemaCollection2_AddRef(pcopy->schemaCache);
         pcopy->XPath = properties->XPath;
+        pcopy->normalize_attributes = properties->normalize_attributes;
+        pcopy->max_element_depth = properties->max_element_depth;
         pcopy->selectNsStr_len = properties->selectNsStr_len;
         list_init( &pcopy->selectNsList );
         pcopy->selectNsStr = malloc(len);
@@ -3086,6 +3092,16 @@
     return IXMLDOMDocument3_validateNode(iface, (IXMLDOMNode*)iface, err);
 }
 
+/* the version 6 document keeps NormalizeAttributeValues and MaxElementDepth */
+static inline BOOL props_stored(const domdoc *This)
+{
+#ifdef SG_MUTANT_PROPS_IGNORED
+    return FALSE;
+#else
+    return This->properties->version == MSXML6;
+#endif
+}
+
 static HRESULT WINAPI domdoc_setProperty(
     IXMLDOMDocument3* iface,
     BSTR p,
@@ -3264,6 +3280,27 @@
             return S_OK;
         }
     }
+    else if (props_stored(This) && (lstrcmpiW(p, PropertyNormalizeAttributeValuesW) == 0 ||
+             lstrcmpiW(p, PropertyMaxElementDepth) == 0))
+    {
+        VARIANT converted;
+        HRESULT hr;
+
+        VariantInit(&converted);
+        if (!lstrcmpiW(p, PropertyNormalizeAttributeValuesW))
+        {
+            if (FAILED(hr = VariantChangeType(&converted, &value, 0, VT_BOOL)))
+                return hr;
+            This->properties->normalize_attributes = V_BOOL(&converted);
+        }
+        else
+        {
+            if (FAILED(hr = VariantChangeType(&converted, &value, 0, VT_I4)))
+                return hr;
+            This->properties->max_element_depth = V_I4(&converted);
+        }
+        return S_OK;
+    }
     else if (lstrcmpiW(p, PropertyProhibitDTDW) == 0 ||
              lstrcmpiW(p, PropertyNewParserW) == 0 ||
              lstrcmpiW(p, PropertyResolveExternalsW) == 0 ||
@@ -3347,6 +3384,18 @@
             return S_OK;
         }
     }
+    else if (props_stored(This) && lstrcmpiW(p, PropertyNormalizeAttributeValuesW) == 0)
+    {
+        V_VT(var) = VT_BOOL;
+        V_BOOL(var) = This->properties->normalize_attributes;
+        return S_OK;
+    }
+    else if (props_stored(This) && lstrcmpiW(p, PropertyMaxElementDepth) == 0)
+    {
+        V_VT(var) = VT_I4;
+        V_I4(var) = This->properties->max_element_depth;
+        return S_OK;
+    }
 
     FIXME("Unknown property %s\n", debugstr_w(p));
     return E_FAIL;
--- a/dlls/msxml3/nodemap.c
+++ b/dlls/msxml3/nodemap.c
@@ -380,11 +380,38 @@
         VARIANT *res, EXCEPINFO *ei)
 {
     xmlnodemap *This = impl_from_IXMLDOMNamedNodeMap( (IXMLDOMNamedNodeMap*)iface );
+    IXMLDOMNode *item = NULL;
 
     TRACE("%p, %ld, %lx, %x, %p, %p, %p.\n", iface, id, lcid, flags, params, res, ei);
 
-    V_VT(res) = VT_DISPATCH;
-    V_DISPATCH(res) = NULL;
+    /* the result is empty, whatever happens */
+#ifndef SG_MUTANT_DISPEX_RESULT
+    if (res) memset(res, 0, sizeof(*res));
+#endif
+
+    if (id == DISPID_VALUE)
+    {
+        VARIANT index;
+        HRESULT hr;
+
+#ifndef SG_MUTANT_NODEMAP_VALUE
+        if (flags != DISPATCH_METHOD && flags != DISPATCH_PROPERTYGET && flags != (DISPATCH_METHOD | DISPATCH_PROPERTYGET))
+            return DISP_E_UNKNOWNNAME;
+        if (params->cArgs - params->cNamedArgs != 1)
+            return DISP_E_BADPARAMCOUNT;
+
+        VariantInit(&index);
+        if (FAILED(hr = VariantChangeType(&index, params->rgvarg, 0, VT_I4)))
+            return hr;
+
+        IXMLDOMNamedNodeMap_get_item(&This->IXMLDOMNamedNodeMap_iface, V_I4(&index), &item);
+        V_VT(res) = VT_DISPATCH;
+        V_DISPATCH(res) = (IDispatch *)item;
+        return S_OK;
+#else
+        return DISP_E_UNKNOWNNAME;
+#endif
+    }
 
     if (id < DISPID_DOM_COLLECTION_BASE || id > DISPID_DOM_COLLECTION_MAX)
         return DISP_E_UNKNOWNNAME;
@@ -393,15 +420,16 @@
     {
         case INVOKE_PROPERTYGET:
         {
-            IXMLDOMNode *disp = NULL;
-
-            IXMLDOMNamedNodeMap_get_item(&This->IXMLDOMNamedNodeMap_iface, id - DISPID_DOM_COLLECTION_BASE, &disp);
-            V_DISPATCH(res) = (IDispatch*)disp;
+            V_VT(res) = VT_DISPATCH;
+            IXMLDOMNamedNodeMap_get_item(&This->IXMLDOMNamedNodeMap_iface, id - DISPID_DOM_COLLECTION_BASE, &item);
+            V_DISPATCH(res) = (IDispatch*)item;
             break;
         }
         default:
         {
             FIXME("unimplemented flags %x\n", flags);
+            V_VT(res) = VT_DISPATCH;
+            V_DISPATCH(res) = NULL;
             break;
         }
     }
--- a/dlls/msxml3/nodelist.c
+++ b/dlls/msxml3/nodelist.c
@@ -324,6 +324,11 @@
 
     TRACE("%p, %ld, %lx, %x, %p, %p, %p.\n", iface, id, lcid, flags, params, res, ei);
 
+    /* the result is empty, whatever happens */
+#ifndef SG_MUTANT_DISPEX_RESULT
+    if (res) memset(res, 0, sizeof(*res));
+#endif
+
     if (id >= DISPID_DOM_COLLECTION_BASE && id <= DISPID_DOM_COLLECTION_MAX)
     {
         switch(flags)
--- a/dlls/msxml3/tests/domdoc.c
+++ b/dlls/msxml3/tests/domdoc.c
@@ -11770,7 +11770,6 @@
     hr = IDispatchEx_Invoke(dispex, DISPID_VALUE, &IID_NULL, 0, DISPATCH_METHOD, &dispparams, &ret, NULL, NULL);
     ok(hr == DISP_E_BADPARAMCOUNT, "Unexpected hr %#lx.\n", hr);
     ok(V_VT(&ret) == VT_EMPTY, "got %d\n", V_VT(&ret));
-    todo_wine
     ok(broken(V_DISPATCH(&ret) == (void*)0x1) || (V_DISPATCH(&ret) == NULL), "got %p\n", V_DISPATCH(&ret));
 
     V_VT(&arg) = VT_I4;
@@ -11785,7 +11784,6 @@
     hr = IDispatchEx_Invoke(dispex, DISPID_VALUE, &IID_NULL, 0, DISPATCH_METHOD, &dispparams, &ret, NULL, NULL);
     ok(hr == DISP_E_BADPARAMCOUNT, "Unexpected hr %#lx.\n", hr);
     ok(V_VT(&ret) == VT_EMPTY, "got %d\n", V_VT(&ret));
-    todo_wine
     ok(broken(V_DISPATCH(&ret) == (void*)0x1) || (V_DISPATCH(&ret) == NULL), "got %p\n", V_DISPATCH(&ret));
 
     V_VT(&arg) = VT_I4;
@@ -11838,7 +11836,6 @@
     hr = IDispatchEx_Invoke(dispex, DISPID_DOM_NODELIST_LENGTH, &IID_NULL, 0, DISPATCH_METHOD, &dispparams, &ret, NULL, NULL);
     ok(hr == DISP_E_MEMBERNOTFOUND, "Unexpected hr %#lx.\n", hr);
     ok(V_VT(&ret) == VT_EMPTY, "got %d\n", V_VT(&ret));
-    todo_wine
     ok(broken(V_I4(&ret) == 1) || (V_I4(&ret) == 0), "got %ld\n", V_I4(&ret));
 
     IXMLDOMNodeList_Release(node_list);
@@ -11937,10 +11934,8 @@
     V_VT(&ret) = VT_EMPTY;
     V_DISPATCH(&ret) = (void*)0x1;
     hr = IDispatchEx_Invoke(dispex, DISPID_VALUE, &IID_NULL, 0, DISPATCH_METHOD, &dispparams, &ret, NULL, NULL);
-todo_wine {
     ok(hr == DISP_E_BADPARAMCOUNT, "Unexpected hr %#lx.\n", hr);
     ok(V_VT(&ret) == VT_EMPTY, "got %d\n", V_VT(&ret));
-}
     ok(broken(V_DISPATCH(&ret) == (void*)0x1) || (V_DISPATCH(&ret) == NULL), "got %p\n", V_DISPATCH(&ret));
 
     V_VT(&arg) = VT_I4;
@@ -11953,10 +11948,8 @@
     V_VT(&ret) = VT_EMPTY;
     V_DISPATCH(&ret) = (void*)0x1;
     hr = IDispatchEx_Invoke(dispex, DISPID_VALUE, &IID_NULL, 0, DISPATCH_METHOD, &dispparams, &ret, NULL, NULL);
-todo_wine {
     ok(hr == DISP_E_BADPARAMCOUNT, "Unexpected hr %#lx.\n", hr);
     ok(V_VT(&ret) == VT_EMPTY, "got %d\n", V_VT(&ret));
-}
     ok(broken(V_DISPATCH(&ret) == (void*)0x1) || (V_DISPATCH(&ret) == NULL), "got %p\n", V_DISPATCH(&ret));
 
     V_VT(&arg) = VT_I4;
@@ -11969,7 +11962,6 @@
     V_VT(&ret) = VT_EMPTY;
     V_DISPATCH(&ret) = (void*)0x1;
     hr = IDispatchEx_Invoke(dispex, DISPID_VALUE, &IID_NULL, 0, DISPATCH_METHOD, &dispparams, &ret, NULL, NULL);
-    todo_wine
     ok(hr == S_OK, "Unexpected hr %#lx.\n", hr);
     ok(V_VT(&ret) == VT_DISPATCH, "got %d\n", V_VT(&ret));
     ok(V_DISPATCH(&ret) == NULL, "got %p\n", V_DISPATCH(&ret));
@@ -12009,7 +12001,6 @@
     V_I4(&ret) = 1;
     hr = IDispatchEx_Invoke(dispex, DISPID_DOM_NODELIST_LENGTH, &IID_NULL, 0, DISPATCH_METHOD, &dispparams, &ret, NULL, NULL);
     ok(hr == DISP_E_MEMBERNOTFOUND, "Unexpected hr %#lx.\n", hr);
-    todo_wine
     ok(V_VT(&ret) == VT_EMPTY, "got %d\n", V_VT(&ret));
     ok(broken(V_I4(&ret) == 1) || (V_I4(&ret) == 0), "got %ld\n", V_I4(&ret));
 
@@ -13694,11 +13685,9 @@
     V_VT(&var) = VT_I2;
     V_I2(&var) = 10;
     hr = IXMLDOMDocument2_getProperty(doc, _bstr_("NormalizeAttributeValues"), &var);
-todo_wine {
     ok(hr == S_OK, "Failed to get property value, hr %#lx.\n", hr);
     ok(V_VT(&var) == VT_BOOL, "Unexpected property value type, vt %d.\n", V_VT(&var));
     ok(V_BOOL(&var) == VARIANT_FALSE, "Unexpected property value.\n");
-}
     V_VT(&var) = VT_BOOL;
     V_BOOL(&var) = VARIANT_TRUE;
     hr = IXMLDOMDocument2_setProperty(doc, _bstr_("NormalizeAttributeValues"), var);
@@ -13707,11 +13696,9 @@
     V_VT(&var) = VT_I2;
     V_I2(&var) = 10;
     hr = IXMLDOMDocument2_getProperty(doc, _bstr_("NormalizeAttributeValues"), &var);
-todo_wine {
     ok(hr == S_OK, "Failed to get property value, hr %#lx.\n", hr);
     ok(V_VT(&var) == VT_BOOL, "Unexpected property value type, vt %d.\n", V_VT(&var));
     ok(V_BOOL(&var) == VARIANT_TRUE, "Unexpected property value.\n");
-}
     IXMLDOMDocument2_Release(doc);
 }
 
@@ -13727,11 +13714,9 @@
     V_VT(&var) = VT_UI4;
     V_UI4(&var) = 0xdeadbeef;
     hr = IXMLDOMDocument2_getProperty(doc, _bstr_("MaxElementDepth"), &var);
-todo_wine {
     ok(hr == S_OK, "Failed to get property value, hr %#lx.\n", hr);
     ok(V_VT(&var) == VT_I4, "Unexpected property value type, vt %d.\n", V_VT(&var));
     ok(V_I4(&var) == 256, "Unexpected property value.\n");
-}
 
     /* Changes to the depth value should be observable when subsequently retrieved. */
     V_VT(&var) = VT_I4;
@@ -13742,11 +13727,9 @@
     V_VT(&var) = VT_UI4;
     V_UI4(&var) = 0xdeadbeef;
     hr = IXMLDOMDocument2_getProperty(doc, _bstr_("MaxElementDepth"), &var);
-todo_wine {
     ok(hr == S_OK, "Failed to get property value, hr %#lx.\n", hr);
     ok(V_VT(&var) == VT_I4, "Unexpected property value type, vt %d.\n", V_VT(&var));
     ok(V_I4(&var) == 32, "Unexpected property value.\n");
-}
 
     IXMLDOMDocument2_Release(doc);
 }
